plugins/usbdmx/AsyncUsbReceiver.cpp

```diff
@@ -0,0 +1,181 @@
+  m_inited_with_handle = true;
+  return true;
+}
+
+bool AsyncUsbReceiver::Start() {
+  if (!m_usb_handle) {
+    OLA_WARN << "AsyncUsbReceiver hasn't been initialized";
+    return false;
+  }
+  ola::thread::MutexLocker locker(&m_mutex);
+  return PerformTransfer();
+}
+
+void AsyncUsbReceiver::CancelTransfer() {
+  if (!m_transfer) {
+    return;
+  }
+
+  bool canceled = false;
+  while (1) {
+    ola::thread::MutexLocker locker(&m_mutex);
+    if (m_transfer_state == IDLE || m_transfer_state == DISCONNECTED) {
+      break;
+    }
+    if (!canceled) {
+      m_suppress_continuation = true;
+      if (m_adaptor->CancelTransfer(m_transfer) == 0) {
+        canceled = true;
+      } else {
+        break;
+      }
+    }
+  }
+
+  m_suppress_continuation = false;
+}
+
+void AsyncUsbReceiver::FillControlTransfer(unsigned char *buffer,
+                                           unsigned int timeout) {
+  m_adaptor->FillControlTransfer(m_transfer, m_usb_handle, buffer,
+                                 &AsyncCallback, this, timeout);
+}
+
+void AsyncUsbReceiver::FillBulkTransfer(unsigned char endpoint,
+                                        unsigned char *buffer,
+                                        int length,
+                                        unsigned int timeout) {
+  m_adaptor->FillBulkTransfer(m_transfer, m_usb_handle, endpoint, buffer,
+                              length, &AsyncCallback, this, timeout);
+}
+
+void AsyncUsbReceiver::FillInterruptTransfer(unsigned char endpoint,
+                                             unsigned char *buffer,
+                                             int length,
+                                             unsigned int timeout) {
+  m_adaptor->FillInterruptTransfer(m_transfer, m_usb_handle, endpoint, buffer,
+                                   length, &AsyncCallback, this, timeout);
+}
+
+int AsyncUsbReceiver::SubmitTransfer() {
+  int ret = m_adaptor->SubmitTransfer(m_transfer);
+  if (ret) {
+    OLA_WARN << "libusb_submit_transfer returned "
+             << m_adaptor->ErrorCodeToString(ret);
+    if (ret == LIBUSB_ERROR_NO_DEVICE) {
+      m_transfer_state = DISCONNECTED;
+    }
+    return false;
+  }
+  m_transfer_state = IN_PROGRESS;
+  return ret;
+}
+
+void AsyncUsbReceiver::TransferComplete(struct libusb_transfer *transfer) {
+  if (transfer != m_transfer) {
+    OLA_WARN << "Mismatched libusb transfer: " << transfer << " != "
+             << m_transfer;
+    return;
+  }
+
+  if (transfer->status != LIBUSB_TRANSFER_COMPLETED &&
+      transfer->status != LIBUSB_TRANSFER_TIMED_OUT ) {
+    OLA_WARN << "Transfer returned " << transfer->status;
+  }
+
+  ola::thread::MutexLocker locker(&m_mutex);
+  m_transfer_state = (transfer->status == LIBUSB_TRANSFER_NO_DEVICE ?
+      DISCONNECTED : IDLE);
+
+  if (m_suppress_continuation) {
+    return;
+  }
+
+  if (transfer->status != LIBUSB_TRANSFER_TIMED_OUT) {
+    if (TransferCompleted(&m_rx_buffer)) {  // Input changed
```

Comment:
You should pass in the actual transfer size to TransferComplete() , in case the USB device sends less than the DATABLOCK size.